### PR TITLE
implement `Copilot` based on the proposed `languageModels` API

### PR DESCRIPTION
**Pull request by wangmingliang-ms**
1. upgrade engine to use LogOutputChannel
2. implement Copilot which encapsulate `llm` API and handles input/ouput(e.g. incomplete response) of `llm`

### CODE REVIEW

File: src/copilot/Copilot.ts [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+            logger.info(`Copilot: \n`, rawAnswer);
+            answer += rawAnswer;
+            return answer.trim().endsWith(this.endMark);
+        };
+        let complete: boolean = await _send(userMessage);
+        while (!complete && rounds < 10) {
+            complete = await _send('continue where you left off.');
+        }
+        logger.debug('rounds', rounds);
+        addContextProperty('rounds', rounds + '');
```
Comment (testforstephen) [inline src/copilot/Copilot.ts:53]: this is a global context property that will be set for all "info" events, and it looks to be misused for me. Why not send an info event directly with the "rounds" data?

**Review by testforstephen — CHANGES_REQUESTED** [review]

File: src/copilot/Copilot.ts (hunk 1 above)
Comment (wangmingliang-ms) [reply src/copilot/Copilot.ts:53]: I had thought it adds property to current telemetry event. I'll fix it.

File: src/copilot/Copilot.ts [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+                logger.error(`Failed to send request to copilot`, e);
+                throw new Error(`Failed to send request to copilot: ${e}`);
+            }
+            messages.push(new LanguageModelChatAssistantMessage(rawAnswer));
+            logger.info(`Copilot: \n`, rawAnswer);
+            answer += rawAnswer;
+            return answer.trim().endsWith(this.endMark);
+        };
+        let complete: boolean = await _send(userMessage);
+        while (!complete && rounds < 10) {
```
[the remark this answers is not in the document]
Comment (wangmingliang-ms) [reply src/copilot/Copilot.ts]: using constant

File: package.json [hunk 3]
```diff
@@ -5,9 +5,9 @@
   "license": "MIT",
   "version": "0.26.0",
   "publisher": "vscjava",
-  "preview": false,
+  "preview": true,
```
[the remark this answers is not in the document]
Comment (wangmingliang-ms) [reply package.json]: reverted.

File: src/copilot/Copilot.ts (hunk 1 above)
Comment (wangmingliang-ms) [reply src/copilot/Copilot.ts]: seems not able to add property to current telemetry operation event. switch to sending another telemetry event.

File: src/copilot/utils.ts [hunk 4]
```diff
@@ -0,0 +1,3 @@
+import { LogOutputChannel, window } from "vscode";
+
+export const logger: LogOutputChannel = window.createOutputChannel("Rewriting Suggestions", { log: true });
```
[the remark this answers is not in the document]
Comment (wangmingliang-ms) [reply src/copilot/utils.ts]: make sense, renamed

**Review by testforstephen — APPROVED** [review]

**Pull request merged**